### PR TITLE
Optimize the performance of the memory pool.

### PR DESCRIPTION
**Pull request by caryliu1999**
https://github.com/cocos-creator/3d-tasks/issues/6760

### CODE REVIEW

File: cocos/bindings/dop/BufferAllocator.cpp [hunk 1]
```diff
@@ -29,19 +29,21 @@
...
+    }
 }
 
 BufferAllocator::~BufferAllocator() {
     for (auto buffer : _buffers) {
         buffer.second->decRef();
     }
     _buffers.clear();
-    BufferAllocator::_pools.erase(_type);
+    _caches.clear();
```
Comment (minggo) [inline cocos/bindings/dop/BufferAllocator.cpp:46]: I think _buffers.clear() and  _caches.clear() is not needed in destructor.

File: cocos/bindings/dop/BufferAllocator.h [hunk 2]
```diff
@@ -37,42 +37,58 @@ namespace se {
...
     ~BufferAllocator();
 
     Object *alloc(uint index, uint bytes);
-    void free(uint index);
+    void    free(uint index);
 
 private:
-    static cc::map<PoolType, BufferAllocator *> _pools;
-    static constexpr uint _bufferMask = ~(1 << 30);
+    static cc::vector<BufferAllocator *> _pools;
```
Comment (minggo) [inline cocos/bindings/dop/BufferAllocator.h:86]: Without pool type information, how to find correct pool?

File: cocos/bindings/dop/BufferAllocator.cpp (hunk 1 above)
Comment (caryliu1999) [reply cocos/bindings/dop/BufferAllocator.cpp:46]: > I think _buffers.clear() and _caches.clear() is not needed in destructor.

Ok, I delete them.

File: cocos/bindings/dop/BufferAllocator.h (hunk 2 above)
Comment (caryliu1999) [reply cocos/bindings/dop/BufferAllocator.h:86]: > Without pool type information, how to find correct pool?

PoolType is continuous, and several different pools are indexed from 0 and then converted to id values based on type.

**Pull request merged**